### PR TITLE
[serve] Fall back to all replicas after those with least multiplexed models

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

As described in https://github.com/ray-project/ray/issues/43965, in the multiplexing path we currently fall back to replicas that have the least number of models loaded. However, this process is deterministic, so if those replicas cannot accept the requests, we will infinitely be in backoff trying to schedule to them.

## Related issue number

https://github.com/ray-project/ray/issues/43965 (closed after cherry pick)

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/tests/unit/test_pow_2_replica_scheduler.py [hunk 1]
```diff
@@ -1084,6 +1084,23 @@ async def test_choose_least_number_of_models_replicas(self, pow_2_scheduler):
             task = loop.create_task(s.choose_replica_for_request(request))
             assert (await task) == r2
 
+    async def test_backoff_from_least_number_of_models_replicas(self, pow_2_scheduler):
```
Comment (edoakes) [inline python/ray/serve/tests/unit/test_pow_2_replica_scheduler.py:1087]: This test hangs forever without the changes in `pow_2_scheduler.py`

Comment (edoakes) [reply python/ray/serve/tests/unit/test_pow_2_replica_scheduler.py:1087]: (and obviously passes now :))

File: python/ray/serve/_private/replica_scheduler/pow_2_scheduler.py [hunk 2]
```diff
@@ -336,15 +312,18 @@ async def choose_two_replicas_with_backoff(
...
+            tried_same_node = False
+
             multiplexed_start_matching_time = None
             multiplexed_matching_timeout = random.uniform(
                 RAY_SERVE_MULTIPLEXED_MODEL_ID_MATCHING_TIMEOUT_S,
                 RAY_SERVE_MULTIPLEXED_MODEL_ID_MATCHING_TIMEOUT_S * 2,
             )
-            tried_same_node = False
-            tried_same_az = False
+            tried_least_multiplexed_models = False
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/replica_scheduler/pow_2_scheduler.py:326]:
```suggestion
            tried_fewest_multiplexed_models = False
```

File: python/ray/serve/_private/replica_scheduler/pow_2_scheduler.py [hunk 3]
```diff
@@ -376,37 +355,40 @@ async def choose_two_replicas_with_backoff(
...
+                        elif candidate_replica_ids:
                             self._multiplexed_model_id_fallback_match.discard(
                                 request_metadata.multiplexed_model_id
                             )
-                    else:
+                    elif not tried_least_multiplexed_models:
+                        # After the `multiplexed_matching_timeout` is up, first try
+                        # routing to replicas that have the fewest models loaded.
+                        # We only try this once to avoid deterministically retrying on
+                        # the same models repeatedly.
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/replica_scheduler/pow_2_scheduler.py:383]:
```suggestion
                        # the same replicas repeatedly.
```

**Review by shrekris-anyscale — APPROVED** [review]
Nice work!

**Pull request merged**